### ISSUE TITLE
Naming

### ISSUE THREAD

**Issue by moretea** [bobvanluijt/weaviate-graphql-prototype#22]
It is important to name things properly, with clear and self describing names.

If I look at the following query, I do not have an idea what it is supposed to do.
```graphql
Local {
   ConvertedFetch(_filter: { ....}) {
     Things {
        NetflixSubscription { level }
      }
   }
}
```

If we would rename ConvertedFetch to Search, and `_filter` to `filter`, or even `where`, we'd get something like this:
```graphql
Local {
   Search(where: { ....} ) {
     Things {
        NetflixSubscription { level }
      }
   }
}
```

**Comment by bobvanluijt** [comment]
First of all, I agree with your statement of proper naming; this does not only go for the functions but also object descriptions which are defined [here](https://github.com/bobvanluijt/weaviate-graphql-prototype/wiki/Schema-definitions). Proper nomenclature is key in developer UX.

Before I dive into some reflection on your suggestions, an overview of the status quo;

# Current Convention

1. One makes a distinction between requests on the network and local requests.
2. The functions always follow the following principle by default: `Location => Function => Graph`
3. Fetch(ing) points to collecting something from the graph. One "fetches a collection" and "retrieves" it. More about this below (**¶ Adjectives**).
4. Fetching is a verb, prefixed with an adjective (meta, fuzzy, converted, etc.). Meaning: "The fetch is meta," "the fetch is fuzzy," etc. This is also described [here](https://github.com/bobvanluijt/weaviate-graphql-prototype/wiki/Schema-definitions). For example: `WeaviateLocalConvertedFetchActionsObj` = _Fetch Actions [from] the internal Weaviate_.

### Core functions

```graphql
{
  Local{
    ConvertedFetch
    HelpersFetch
    MetaFetch
  }
  Network{
    MetaFetch
    FuzzyFetch
    ConvertedFetch
  }
}
```

### Adjectives

- **Fetching** implies that some entity needs to go and get something that is remote and bring it back. i.e.: Dogs fetch a stick.
- **Getting** implies that you have sole possession of something.
- **Searching** implies you are thoroughly scrutinizing.
- **Filtering** implies you remove unwanted material.
- **Generics** implies something that relates to or characteristic of a whole group or class.
- **Analyzing** implies that one studies or determines the nature and relationship of the parts of (something).

_Note: `Statistics` is a noun, therefore I'll skip this one_

### Usage of underscores

- 2 underscores = graphql reserved.
- 1 underscore = none Weaviate class or property
- 0 underscores = Weaviate class or property

_Note: this was created to avoid confusion in classes or properties vs. attributes. However, we don't use none-underscored attributes._

### Usage of _where_

I find this very powerful to make the queries more semantically attractive.

# Conclusion and reflection

Per level;

### Level 1

Choosing if one wants to execute a request over the `Network` or `Local` seem correct.

### Level 2

- Over the network we are _fetching_ and _retrieving_. I.e. the Weaviate "fetches" something over the network. However, it does not fetch anything local. It _gets_ something local.
- Instead of `search` I would suggest `get`. The argument `where` is compelling, and I would keep this in.
- I agree with the point about `Generics,` why do we have this again? Why was this not part of the `MetaFetch`? (cc: @laura-ham)
- I do think that "meta" still describes that we are getting "metadata" about the things or actions we are getting.
- The underscores don't make any sense because we don't use classes or properties, removing these makes complete sense.

# Suggestion

### Core functions

```graphql
{
  Local{
    Get
    GetHelpers
    Meta
  }
  Network{
    Fetch
    Meta
  }
}
```

### Where

```graphql
{
  Local{
    Get (where:{ # Directly from Get function, needs to be validated.
        EQ: [{
          path: ["Things", "City", "inCountry", "Country", "name"],
          value: "Netherlands"
        }]
      }){
# ...
```

**Comment by laura-ham** [comment]
I agree; my comments on this:

**Level 1.** 
Let's keep layer 1 with the distinction between Network and Local

**Level 2.** 
- Let's have 'Get' as you proposed. I'm not sure about GetHelpers, but since we are not using this query type at this moment I suggest to keep it our for now and leave the Local with Get and Meta. When we define a use case and specify the query design for the 'GetHelpers' we should find an appropriate name for that.
- 'where' instead of '_filters' is a good suggestion, I agree with the explanation
- I think the 'Generics' in 'Metafetch' makes indeed no sense. Maybe we should consider still splitting up the meta query, because there is a difference in querying metadata about an ontology/schema (types etc), and the meta statistics about the actual data in the individual nodes (e.g. min/max/average number). (see https://en.wikipedia.org/wiki/Metadata for different types of metadata.). The only doubt about (sub)dividing the Meta query into multiple (statistical, structural) is that it may make it more difficult for the user to receive information in a simple query.
- I'll remove the underscores

**Comment by moretea** [comment]
@bobvanluijt thats fine! I have no strong opinions about the actual words that are used, just that the final queries are self-explaining.

Most people that are going to using the API will be programmers that are not experts on the semantic interpretation of words. For them, a query should that takes less than two seconds to understand.

One thing that I think we _should_ avoid, is using different names for local and network queries. For now we can continue using different names as you suggested (Get for local, and Fetch for network), but before we start implementing anything, let's evaluate how the queries actually would look like.

The reason for that is that IMHO it is very valuable if you can write a  local query, and then just flip from local to network, and get remote results.

**Comment by bobvanluijt** [comment]
Hi @laura-ham, @moretea,

Based on the discussion above I would like to cut the knot to the following next steps.

1. The second step (in Local and Network)
```
{
  Local{
    Get
    Meta
  }
  Network{
    Fetch
    Meta
  }
}
```
2. We use the following rules: if a second step function also exists in the opposite root function (for example in Local or in Network) the function should be identical. Otherwise, we should use clear nomenclature and descriptions to explain the difference to end users. The goal of this is that one can "just flip from local to network, and get remote results."
3. Remove underscores.
4. Implement `where` clause.
5. Implement improved filters as discussed in this thread: https://github.com/creativesoftwarefdn/weaviate/issues/441

@laura-ham, that makes it an action for you first, later for @JeroenStravers1

**Comment by laura-ham** [comment]
@bobvanluijt, point 2 is a great idea, but the filters and results will always be different for 'meta' in your example. This is because we want to filter by and give results about 'certainty' and 'keywords'. For now, I designed the first two levels like: 

``` graphql
{
  Local{
    Get
    GetMeta
  }
  Network{
    Fetch # Fetch things/actions in a fuzzy way
    Introspect # discover what's in the ontologies over the network
    FetchMeta # get meta information like counts and averages? (no design yet)
  }
}
```

**Comment by bobvanluijt** [comment]
That's okay, now they are clearly separated.
